### PR TITLE
[data] Moving collate: enable fusion for streaming_repartition

### PR DESCRIPTION
**Pull request by xinyuangui2**
## Summary

It will become a common pattern for
```
ds.repartition(target_num_rows_per_block=batch_size).map_batches(collate_fn, batch_size=batch_size)
```
Right now these 2 operators are not fused, which will cost one extra serialization and deserialization cost. We want to put `streaming_repartition` into the operator fusion as well.

https://github.com/ray-project/ray/issues/54431 is not a blocker to the streaming_repartition fusion anymore.

### RefBundler Fusion

RefBundler fusion handles how input `refbundle`s are grouped before processing. We start from minimum extra fusion.

- **`StreamingRepartitionRefbundler(target_num_rows=X)`**  
  Ensures each input refbundle has a size that is a multiple of `target_num_rows`.

- **`BlockRefbundler(min_num_rows=Y)`**  
  Ensures each input refbundle is at least `min_num_rows`.

We implemented one naive fusion strategy

1. **StreamingRepartition ⨯ StreamingRepartition**  
   Two `StreamingRepartitionRefbundler`s can be merged only if they share the same `target_num_rows`.  
   - Result: the same `StreamingRepartitionRefbundler`.

2. **StreamingRepartition ⨯ BlockRefbundler**  (not the opposite direction)
   A `StreamingRepartitionRefbundler(target_num_rows=X)` can merge with a `BlockRefbundler(min_num_rows=Y)` if `X == Y` or `Y is None`.
   - Result: `StreamingRepartitionRefbundler(target_num_rows=X)`.

3. **BlockRefbundler ⨯ BlockRefbundler**  
   Two `BlockRefbundler`s always merge.  
   - Result: `BlockRefbundler(min_num_rows=max(X, Y))`.

## Test Result

Text tokenization script: https://gist.github.com/xinyuangui2/211d364baeec8ddd96cb7ae81df1ca5b

With fusion, we are able to see obvious speed advantage over no fusion.

## Issue

https://github.com/ray-project/ray/issues/58837

### CODE REVIEW

File: python/ray/data/_internal/planner/plan_udf_map_op.py [hunk 1]
```diff
@@ -156,11 +156,12 @@ def plan_streaming_repartition_op(
     assert len(physical_children) == 1
     input_physical_dag = physical_children[0]
     compute = get_compute(op._compute)
-    transform_fn = BlockMapTransformFn(
-        lambda blocks, ctx: blocks,
-        output_block_size_option=OutputBlockSizeOption.of(
```
Comment (xinyuangui2) [inline python/ray/data/_internal/planner/plan_udf_map_op.py:161]: `output_block_size_option` is likely to be override, so we cannot rely on it.

File: python/ray/air/_internal/torch_utils.py [hunk 2]
```diff
@@ -436,6 +436,12 @@ def concat_tensors_to_device(
         [type(t) for t in tensor_sequence if not isinstance(t, torch.Tensor)]
     )
 
+    # If there is only one tensor and its device already matches, return it directly.
+    if len(tensor_sequence) == 1 and (
+        device is None or tensor_sequence[0].device == torch.device(device)
+    ):
+        return tensor_sequence[0]
```
[the remark this answers is not in the document]
Comment (xinyuangui2) [reply python/ray/air/_internal/torch_utils.py:511]: Should be fine.

File: python/ray/data/_internal/logical/rules/operator_fusion.py [hunk 3]
```diff
@@ -505,6 +527,84 @@ def _can_fuse_map_ops(
...
+                    target_num_rows_per_block=up_ref_bundler._target_num_rows
+                )
+            else:
+                # TODO(xgui): Explore if we can use least common multiple of the two target_num_rows_per_block
+                return None
+        else:
+            supported_types = (BlockRefBundler, StreamingRepartitionRefBundler)
+            assert isinstance(up_ref_bundler, supported_types) and isinstance(
+                down_ref_bundler, supported_types
+            )
```
Comment (raulchen) [inline python/ray/data/_internal/logical/rules/operator_fusion.py]: we should only fuse BlockRefBundler -> StreamingRepartitionRefBundler, not the opposite.

File: python/ray/data/_internal/planner/plan_udf_map_op.py [hunk 4]
```diff
@@ -156,11 +156,12 @@ def plan_streaming_repartition_op(
...
-            target_num_rows_per_block=op.target_num_rows_per_block,  # To split n*target_max_block_size row into n blocks
+    transform_fn = BatchMapTransformFn(
+        _generate_transform_fn_for_map_batches(
+            lambda blocks: blocks,
         ),
+        batch_size=op.target_num_rows_per_block,
+        disable_block_shaping=True,
```
Comment (raulchen) [inline python/ray/data/_internal/planner/plan_udf_map_op.py:164]: is this needed?
I think `batch_size=op.target_num_rows_per_block` should take care of it.

File: python/ray/data/_internal/streaming_repartition.py [hunk 5]
```diff
@@ -18,12 +18,7 @@
        and add the remaining bundle to the pending bundles for the next iteration.
     4. Submit that ready bundle to a remote map task; the task slices each block according to the slice metadata stored
        in the RefBundle (the bundle now contains n × target rows for n ≥ 1).
-    5. We configured the `OutputBlockSizeOption.target_num_rows_per_block` to the target number of rows per block in
-       plan_streaming_repartition_op so the output buffer further splits the n × target rows into n blocks of exactly
-       the target size.
-       Note: the output buffer only splits a bundle when its row count exceeds `target_rows × MAX_SAFE_ROWS_PER_BLOCK_FACTOR`
-       (default 1.5). Because we split bundles into target-row blocks, `MAX_SAFE_ROWS_PER_BLOCK_FACTOR` must stay < 2 to
-       output the target-row blocks.
+    5. We create one BatchMapTransformFn and its batcher will create blocks with exactly the target number of rows.
```
Comment (raulchen) [inline python/ray/data/_internal/streaming_repartition.py]: BatchMapTransformFn is an implementation detail. better not put it here.

File: python/ray/data/tests/test_operator_fusion.py [hunk 6]
```diff
@@ -742,6 +750,586 @@ def test_zero_copy_fusion_eliminate_build_output_blocks(
...
+            assert isinstance(bundler, BlockRefBundler), (
+                f"{description}\n"
+                f"Operator {op_position} ({op.name}): Expected BlockRefBundler "
+                f"but got {type(bundler).__name__}"
+            )
+            assert bundler._min_rows_per_bundle == expected_value, (
+                f"{description}\n"
+                f"Operator {op_position} ({op.name}): Expected min_rows_per_bundle={expected_value} "
+                f"but got {bundler._min_rows_per_bundle}"
+            )
```
Comment (raulchen) [inline python/ray/data/tests/test_operator_fusion.py]: besides unit tests, let's also add an e2e test - use `map_batches().repartition()` and check the ops are properly fused.

File: python/ray/data/_internal/planner/plan_udf_map_op.py (hunk 4 above)
Comment (xinyuangui2) [reply python/ray/data/_internal/planner/plan_udf_map_op.py:164]: `batch_size=op.target_num_rows_per_block` only makes sure the `lambda blocks, ctx: blocks,` receives the exactly batch_size, but here we want to ensure the output blocks keep the shape.

File: python/ray/air/_internal/torch_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (srinathk10) [reply python/ray/air/_internal/torch_utils.py:511]: Nice! Do you have a test that exercises this?

File: python/ray/data/_internal/logical/rules/operator_fusion.py [hunk 7]
```diff
@@ -505,6 +527,84 @@ def _can_fuse_map_ops(
...
+        elif isinstance(up_ref_bundler, StreamingRepartitionRefBundler) and isinstance(
+            down_ref_bundler, StreamingRepartitionRefBundler
+        ):
+            if up_ref_bundler._target_num_rows == down_ref_bundler._target_num_rows:
+                return StreamingRepartitionRefBundler(
+                    target_num_rows_per_block=up_ref_bundler._target_num_rows
+                )
+            else:
+                # TODO(xgui): Explore if we can use least common multiple of the two target_num_rows_per_block
```
Comment (srinathk10) [inline python/ray/data/_internal/logical/rules/operator_fusion.py:583]: Here we could just always override with downstream _target_num_rows. If there are back-to-back StreamingRepartitionRefBundler, the downstream one can just take precedence.

File: python/ray/data/_internal/logical/rules/operator_fusion.py (hunk 3 above)
Comment (xinyuangui2) [reply python/ray/data/_internal/logical/rules/operator_fusion.py]: Good point. I will only limit `StreamingRepartitionRefBundler -> BlockRefBundler` since that is the common pattern for collate fn.

File: python/ray/data/_internal/logical/rules/operator_fusion.py (hunk 7 above)
Comment (xinyuangui2) [reply python/ray/data/_internal/logical/rules/operator_fusion.py:583]: Here I limit the fusion with exactly the same `_target_num_rows` because I don't want the fusion to drop the parallelism. But we can revisit this in the future,.

File: python/ray/air/_internal/torch_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (xinyuangui2) [reply python/ray/air/_internal/torch_utils.py:511]: Added one test case.

File: python/ray/data/tests/test_operator_fusion.py (hunk 6 above)
Comment (xinyuangui2) [reply python/ray/data/tests/test_operator_fusion.py]: Added.

File: python/ray/data/_internal/logical/rules/operator_fusion.py [hunk 8]
```diff
@@ -505,6 +527,75 @@ def _can_fuse_map_ops(
...
+            A compatible merged BaseRefBundler if the bundlers can be fused together,
+            None if they are incompatible and fusion should not occur.
+
+        Fusion rules:
+            - BlockRefBundler + BlockRefBundler: Always compatible. Returns a
+              BlockRefBundler with min_rows_per_bundle set to the maximum of the two.
+            - StreamingRepartitionRefBundler + StreamingRepartitionRefBundler: Only
+              compatible if both have the same target_num_rows_per_block.
+            - StreamingRepartitionRefBundler + BlockRefBundler: Compatible
+              if target_num_rows >= min_rows_per_bundle (or min_rows_per_bundle is None).
```
Comment (xinyuangui2) [inline python/ray/data/_internal/logical/rules/operator_fusion.py]:
```suggestion
              if target_num_rows == min_rows_per_bundle (or min_rows_per_bundle is None).
```

File: python/ray/data/tests/test_torch_tensor_utils.py [hunk 9]
```diff
@@ -66,5 +69,34 @@ def test_move_invalid_batch_type():
         move_tensors_to_device({"a": torch.ones(1), "b": [torch.ones(1)]}, device)
 
 
+def test_concat_tensors_to_device_no_copy():
```
Comment (srinathk10) [inline python/ray/data/tests/test_torch_tensor_utils.py:72]: Nice

File: python/ray/data/tests/test_repartition_e2e.py [hunk 10]
```diff
@@ -373,6 +353,99 @@ def test_repartition_guarantee_row_num_to_be_exact(
             )
 
 
+def test_streaming_repartition_with_partial_last_block(
+    ray_start_regular_shared_2_cpus, disable_fallback_to_object_extension
+):
+    """Test repartition with target_num_rows_per_block where last block has fewer rows.
+    This test verifies:
+    1. Most blocks have exactly target_num_rows_per_block rows
```
Comment (srinathk10) [inline python/ray/data/tests/test_repartition_e2e.py:361]: Can refine this as,
N-1 Blocks have exactly target_num_rows_per_block
Nth Block has remainder rows.

File: python/ray/data/_internal/logical/rules/operator_fusion.py [hunk 11]
```diff
@@ -505,6 +527,75 @@ def _can_fuse_map_ops(
...
+        Returns:
+            A compatible merged BaseRefBundler if the bundlers can be fused together,
+            None if they are incompatible and fusion should not occur.
+
+        Fusion rules:
+            - BlockRefBundler + BlockRefBundler: Always compatible. Returns a
+              BlockRefBundler with min_rows_per_bundle set to the maximum of the two.
+            - StreamingRepartitionRefBundler + StreamingRepartitionRefBundler: Only
+              compatible if both have the same target_num_rows_per_block.
+            - StreamingRepartitionRefBundler + BlockRefBundler: Compatible
```
Comment (srinathk10) [inline python/ray/data/_internal/logical/rules/operator_fusion.py:553]: Any reason we don't fuse,

map_batches->streaming_repartion case.

We only fuse,

streaming_repartition->map_batches

Comment (xinyuangui2) [reply python/ray/data/_internal/logical/rules/operator_fusion.py:553]: `streaming_repartition->map_batches` will be the pattern that we recommend users using to move collate_function. And we want to limit to that for now to avoid unexpected regression from fusion.

File: python/ray/data/_internal/planner/plan_udf_map_op.py (hunk 4 above)
Comment (raulchen) [reply python/ray/data/_internal/planner/plan_udf_map_op.py:164]: sorry, I mean `target_num_rows_per_block=op.target_num_rows_per_block`, which is exactly what we need - make the output blocks have N rows per block.

File: python/ray/data/_internal/logical/rules/operator_fusion.py (hunk 11 above)
Comment (raulchen) [reply python/ray/data/_internal/logical/rules/operator_fusion.py:553]: wait, we should recommend `map_batches->streaming_repartition`. this ensures the output block size.

Comment (raulchen) [reply python/ray/data/_internal/logical/rules/operator_fusion.py:553]: `streaming_repartition->map_batches(batch_size=None)` can also ensure output block size. 
But the `map_batches->streaming_repartition` is more flexible as users can use whatever batch_size you want
